src/storage/src/hummock/iterator/forward_user.rs

```diff
@@ -239,44 +201,102 @@ impl<I: HummockIterator<Direction = Forward>> UserIterator<I> {
             epoch_with_gap: EpochWithGap::new(self.read_epoch, MAX_SPILL_TIMES),
         };
         self.iterator.seek(full_key).await?;
+        // We check `is_valid` and `user_key_out_of_range` here to avoid `delete_range_iter.seek`
+        // as much as possible because it can be expensive. We can simplify this logic after delete
+        // range is deprecated and replaced with vnode watermark.
         if !self.iterator.is_valid() {
             return Ok(());
         }
 
-        if self.key_out_of_range() {
-            self.out_of_range = true;
+        if self.user_key_out_of_range(self.iterator.key().user_key) {
             return Ok(());
         }
 
         self.delete_range_iter.seek(full_key.user_key).await?;
 
-        // Handle multi-version
-        self.last_key = FullKey::default();
-        // Handle range scan when key > end_key
-
-        self.next().await
+        self.try_advance_to_next_valid().await
     }
 
     /// Indicates whether the iterator can be used.
     pub fn is_valid(&self) -> bool {
-        // Handle range scan
-        // key >= begin_key is guaranteed by seek/rewind function
-        (!self.out_of_range) && self.iterator.is_valid()
+        self.is_current_pos_valid
     }
 
     pub fn collect_local_statistic(&self, stats: &mut StoreLocalStatistic) {
         stats.add(&self.stats);
         self.iterator.collect_local_statistic(stats);
     }
 
+    /// Advance the inner iterator to a valid position, in which the entry can be exposed.
+    /// Iterator will not be advanced if it already pointed to a valid position.
+    async fn try_advance_to_next_valid(&mut self) -> HummockResult<()> {
```

Comment:
These are the codes from another merged PR. It seems that there are some conflicts in the PR branch.

Reply:
Got it, I will resolve the conflicts later.